### ISSUE TITLE
can I add protected different password for my different datasets?

### ISSUE THREAD

**Issue by FADHLyemen** [Novartis/cellxgene-gateway#42]
can I add protected different password for my different datasets?

**Comment by alokito** [comment]
Hi, this is not currently supported. At some point we would like to enable dataset level authorization via a pluggable interface. This should enable both password based authorization as well as identity-based authorization via e.g. IAM or AD roles with the appropriate plugin.

**Comment by alokito** [comment]
Hi @FADHLyemen ! I am finally adding something related to this; however I am just handling the authorization part: https://github.com/Novartis/cellxgene-gateway/wiki/Data-based-Authorization
You will need to add some other mechanism for authentication, such as basic auth via nginx or similar. I realize this may be more work for your use case, but separating the authentication from the authorization makes it easier to compose with enterprise authentication mechanisms such as OIDC Connect. I will likely close this ticket after this feature is release in v0.3.8.

**Comment by alokito** [comment]
This has been released. In case it's not clear, you'll need to override the `is_authorized` method specific to your use case. See the above wiki page for an example.

**Comment by Dane-Z** [comment]
Hello, my lab is trying to follow the wiki to implement this but need a little more information. We want to be able to give researchers (within our VPN) a URL that allows them to navigate to cellxgene-gateway and only see or access their own research data within a certain group directory. Is it possible to do this without having them touch the command line in order to set a cookie for themselves to view the data? We have multiple groups that will be using our self-host cellxgene-gateway server on the web but have no experience with the command line so we're trying to find a streamlined way to give members of a group access to particular data sets

**Comment by alokito** [comment]
Hi Dane! The main issue is: how can we know which group a given researcher belongs to? This is more or less a question of "authentication". Large companies often have "IAM solutions" that address this problem. The old school way to approach this (which works fine in smaller settings) is to set up a separate username/password combination for each group. Then when a researcher wants to view data for a particular group, the browser asks them to enter the username and password corresponding to the group. Such a scheme is called "basic auth". A reasonably competent developer should be able to configure basic auth with an nginx reverse proxy or similar in front of cellxgene-gateway. It is also possible to set this up directly in cellxgene-gateway using a module such as flask basic-auth, though it will require a bit of python coding https://flask-httpauth.readthedocs.io/en/latest/. Do either of these paths make sense to you?

At the end of the basic auth dance (https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication), there will be an "Authorization" header that identifies the user (as a base64 encoded username:password pair). This can be used by the `is_authorized` function to grant access to specific datasets based on the user (really group) name and directory name.
I may be able scrape together some time to put a more fully-featured example online if you can provide more details on what would be useful in your scenario.

**Comment by alokito** [comment]
Is there some way that you can pass your existing system IDs to cellxgene gateway? Sometimes there is an "IAM Gateway" that can add the ID as a header to the request, thus making it available to cellxgene gateway.
```

**Comment by alokito** [comment]
> We've also considered assigning a random port within a range of ports for each group's instance of cellxgene-gateway but would rather avoid that if it's possible to restrict access to directories

That's a "security by obscurity" approach... if you're going to go that route, maybe the simplest thing would be to implement a custom file item source that never lists the top level folders? Basically, the "filter" parameter corresponds to the folder path to be listed. You would just return no files when no filter is specified. The only way to see a listing of another group would be to "guess" the path of it's folder. The code is here:

https://github.com/Novartis/cellxgene-gateway/blob/master/cellxgene_gateway/items/file/fileitem_source.py#L53

**Comment by Dane-Z** [comment]
Obscuring the top level folders was a perfect solution! We ended up using your code for the custom list_items function then commenting out "Navigation: " and hyperlinks leading to "top level" and "Home page" from the filecrawler.html file. On top of that we added a small bash script to create the group directories with a random sequence of letters/numbers so that they cannot be easily guessed by any curious researchers. Thank you so much for the informative and thorough instruction and help!!
